File: services/bg_remove_service.py
import os
from uuid import uuid4
from PIL import Image
from rembg import remove
from fastapi import UploadFile

BASE_DIR = "static/bgremove"
PREVIEW_DIR = os.path.join(BASE_DIR, "preview")
FINAL_DIR = os.path.join(BASE_DIR, "final")

os.makedirs(PREVIEW_DIR, exist_ok=True)
os.makedirs(FINAL_DIR, exist_ok=True)

async def bg_remove_service(
    image: UploadFile,
    bg_type: str,
    bg_color: str | None = None,
    bg_image: UploadFile | None = None,
):
    uid = str(uuid4())

    input_path = os.path.join(BASE_DIR, f"{uid}_input.png")
    preview_path = os.path.join(PREVIEW_DIR, f"{uid}_preview.png")
    final_path = os.path.join(FINAL_DIR, f"{uid}_final.png")

    # save input
    with open(input_path, "wb") as f:
        f.write(await image.read())

    original = Image.open(input_path).convert("RGBA")

    # remove bg
    removed = remove(original)
    removed.save(preview_path)

    final_img = removed

    # solid color bg
    if bg_type == "color" and bg_color:
        bg = Image.new("RGBA", removed.size, bg_color)
        bg.paste(removed, (0, 0), removed)
        final_img = bg

    # image bg
    if bg_type == "image" and bg_image:
        bg_path = os.path.join(BASE_DIR, f"{uid}_bg.png")
        with open(bg_path, "wb") as f:
            f.write(await bg_image.read())

        bg = Image.open(bg_path).convert("RGBA")
        bg = bg.resize(removed.size)
        bg.paste(removed, (0, 0), removed)
        final_img = bg

    final_img.save(final_path)

    return {
        "preview": f"/static/bgremove/preview/{uid}_preview.png",
        "final": f"/static/bgremove/final/{uid}_final.png",
    }

File: src/modules/bg-remove/bgRemove.service.ts
import fs from "fs";
import path from "path";
import sharp from "sharp";
import { v4 as uuid } from "uuid";
import { removeBackground } from "@imgly/background-removal-node";

const BASE_DIR = "static/bgremove";
const PREVIEW_DIR = path.join(BASE_DIR, "preview");
const FINAL_DIR = path.join(BASE_DIR, "final");

fs.mkdirSync(PREVIEW_DIR, { recursive: true });
fs.mkdirSync(FINAL_DIR, { recursive: true });

type BgType = "transparent" | "color" | "image";

export const bgRemoveService = async (
  imagePath: string,
  bgType: BgType,
  bgColor?: string,
  bgImagePath?: string
) => {
  const uid = uuid();

  const previewPath = path.join(PREVIEW_DIR, `${uid}_preview.png`);
  const finalPath = path.join(FINAL_DIR, `${uid}_final.png`);

  // 🔹 Remove background
  const inputBuffer = fs.readFileSync(imagePath);
  const removedBuffer = await removeBackground(inputBuffer);

  // Save preview (transparent)
  await sharp(removedBuffer as any).png().toFile(previewPath);

  let finalImage = sharp(removedBuffer as any);

  // 🔹 Solid color background
  if (bgType === "color" && bgColor) {
    const meta = await sharp(removedBuffer as any).metadata();

    const bg = sharp({
      create: {
        width: meta.width!,
        height: meta.height!,
        channels: 4,
        background: bgColor
      }
    });

    finalImage = bg.composite([
      { input: removedBuffer as any, blend: "over" }
    ]);
  }

  // 🔹 Image background
  if (bgType === "image" && bgImagePath) {
    const fgMeta = await sharp(removedBuffer as any).metadata();

    const bg = sharp(bgImagePath).resize(
      fgMeta.width!,
      fgMeta.height!
    );

    finalImage = bg.composite([
      { input: removedBuffer as any, blend: "over" }
    ]);
  }

  await finalImage.png().toFile(finalPath);

  return {
    preview: `/static/bgremove/preview/${uid}_preview.png`,
    final: `/static/bgremove/final/${uid}_final.png`
  };
};
